csrc/scheduler/normalization_inner_outer.cpp

```diff
@@ -0,0 +1,857 @@
+void InnerOuterPersistentKernelScheduler::computeHeuristics(
+    Fusion* fusion,
+    SchedulerRuntimeInfo& runtime_info,
+    HeuristicSummary* data_cache) {
+  params_ = getPersistentHeuristic(fusion, runtime_info, data_cache);
+  NVF_ERROR(params_ != nullptr);
+}
+
+void InnerOuterPersistentKernelScheduler::schedule(Fusion* fusion) {
+  FUSER_PERF_SCOPE("Schedule InnerOuterPersistent Fusion");
+  schedulePersistentKernel(fusion, reductionParams());
+}
+
+bool InnerOuterPersistentKernelScheduler::canScheduleCompileTime(
+    Fusion* fusion) {
+  // (1) leading common checks for all persistent kernels.
+  if (!normalization_scheduler_utils::checkOpsAndInputs(
+          fusion, schedule_heuristic)) {
+    return false;
+  }
+
+  // (2) check reduction type.
+  const auto& reduction_tvs = scheduler_utils::getReductionTvs(fusion);
+  if (!normalization_scheduler_utils::checkReductionType(
+          reduction_tvs, schedule_heuristic)) {
+    return false;
+  }
+
+  // (3) special check for InnerOuter persistent kernel.
+  std::vector<TensorView*> inner_reduction_tvs;
+  std::vector<TensorView*> outer_reduction_tvs;
+  for (auto tv : reduction_tvs) {
+    if (scheduler_utils::isFastestDimReduction(tv)) {
+      inner_reduction_tvs.emplace_back(tv);
+    } else {
+      outer_reduction_tvs.emplace_back(tv);
+    }
+  }
+  normalization_scheduler_utils::checkReductionAxis(
+      fusion, inner_reduction_tvs, schedule_heuristic);
+  normalization_scheduler_utils::checkReductionAxis(
+      fusion, outer_reduction_tvs, schedule_heuristic);
+  if (!normalization_scheduler_utils::checkIfReductionsAreInnerOuter(
+          inner_reduction_tvs, outer_reduction_tvs)) {
+    scheduler_debug_utils::canScheduleRejectReason(
+        schedule_heuristic,
+        "to use combined reduction, inner reduction tensor should be [I,I,...,R,R] and outer reduction tensor should be [R,R,...,I,I]");
+    return false;
+  }
+
+  if (!normalization_scheduler_utils::hasSharedInput(
+          inner_reduction_tvs, outer_reduction_tvs)) {
+    scheduler_debug_utils::canScheduleRejectReason(
+        schedule_heuristic,
+        "to use combined reduction, inner reduction and outer reduction should have shared input.");
+    return false;
+  }
+
+  if (!normalization_scheduler_utils::isConnectedOnlyThroughReductionProducer(
+          inner_reduction_tvs, outer_reduction_tvs)) {
+    scheduler_debug_utils::canScheduleRejectReason(
+        schedule_heuristic,
+        "to use combined reduction, inner reduction and outer reduction should not have shared consumer, their consumers should not have shared non-outer-reduction producer.");
+    return false;
+  }
+
+  // (4) tailing common checks for all persistent kernels.
+  if (!normalization_scheduler_utils::checkViewRootPersistentTopology(
+          fusion, reduction_tvs, inner_reduction_tvs[0], schedule_heuristic)) {
+    return false;
+  }
+
+  return true;
+}
+
+bool InnerOuterPersistentKernelScheduler::canScheduleRunTime(
```

Comment:
adapted from the original runtime check. Implementations are different for different schedulers.